automation/src/test/java/com/microsoft/identity/common/test/automation/questions/AccessToken.java

```diff
@@ -33,26 +33,40 @@
 import net.serenitybdd.screenplay.Question;
 import net.serenitybdd.screenplay.questions.Text;
 
+import org.apache.http.util.TextUtils;
+
 public class AccessToken implements Question<String> {
+
+    private String clientId;
+    public AccessToken(String clientId){
+        this.clientId = clientId;
+    }
+
     @Override
     public String answeredBy(Actor actor) {
         String results = Text.of(Results.RESULT_FIELD).viewedBy(actor).asString();
         ReadCacheResult readCacheResult = ResultsMapper.GetReadCacheResultFromString(results);
         String accessToken = null;
         if (readCacheResult != null) {
             User user = (User) actor;
-            user.setCacheResult(readCacheResult.tokenCacheItems.get(0));
             for (TokenCacheItemReadResult readResult : readCacheResult.tokenCacheItems) {
                 if (readResult.accessToken != null) {
-                    accessToken = readResult.accessToken;
-                    break;
+                    if (TextUtils.isEmpty(clientId) || (!TextUtils.isEmpty(clientId) && clientId.equals(readResult.clientId))) {
+                        accessToken = readResult.accessToken;
+                        user.setCacheResult(readResult);
```

Comment:
Been meaning to ask about this.  Not sure i'm comfortable with the questions having side affects on user state.  I think i'd prefer that these state changes occur inline in the test.  I came across this when debugging an error recently.... i was trying to figure out what set the cacheResult on the user... and had to go digging.  Perhaps this is something that we can refactor in a separate PR though.

Reply:
Thanks for pointing out. I have refactored the code and remove this piece.